src/proof_rules/calculus/soundness_checker.rs

```diff
@@ -1,121 +1,310 @@
+            .any(|approx| approx.kind == ApproximationKind::OVER)
+            && self
+                .approximations
+                .iter()
+                .any(|approx| approx.kind == ApproximationKind::UNDER);
+
+        if let Some(calculus) = self.calculus {
+            match self.overall_approximation() {
+                ApproximationKind::OVER => Some(format!(
+                    "the {} of the program is over-approximated, i.e., vc[program] ≥ {}[program]",
+                    calculus, calculus
+                )),
+                ApproximationKind::UNDER => Some(format!(
+                    "the {} of the program is under-approximated, i.e., vc[program] ≤ {}[program]",
+                    calculus, calculus
+                )),
+                ApproximationKind::UNKNOWN => {
+                    if is_both_over_and_under {
+                        Some(format!(
+                            "the {} of the program is both over- and under-approximated in different parts of the program, therefore the approximation relationship for the whole program is unknown.",
+                            calculus,
+                        ))
+                    } else {
+                        Some(format!(
+                            "the approximation relationship of the vc and the {} of the program is unknown.",
+                            calculus,
+                        ))
+                    }
+                }
+                _ => None,
+            }
+        } else {
+            None
         }
     }
+
+    fn overall_approximation(&self) -> ApproximationKind {
+        self.approximations.infimum()
+    }
 }
 
-impl PartialOrd for ApproximationKind {
-    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
-        match (self, other) {
-            (ApproximationKind::Exact, ApproximationKind::Exact) => Some(std::cmp::Ordering::Equal),
-            (ApproximationKind::Under, ApproximationKind::Under) => Some(std::cmp::Ordering::Equal),
-            (ApproximationKind::Over, ApproximationKind::Over) => Some(std::cmp::Ordering::Equal),
-            (ApproximationKind::Unknown, ApproximationKind::Unknown) => {
-                Some(std::cmp::Ordering::Equal)
-            }
-            (_, ApproximationKind::Exact) => Some(std::cmp::Ordering::Less),
-            (ApproximationKind::Exact, _) => Some(std::cmp::Ordering::Greater),
-            (ApproximationKind::Unknown, _) => Some(std::cmp::Ordering::Less),
-            (_, ApproximationKind::Unknown) => Some(std::cmp::Ordering::Greater),
-            (ApproximationKind::Under, ApproximationKind::Over)
-            | (ApproximationKind::Over, ApproximationKind::Under) => None,
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
+pub struct ApproximationKind {
+    under: bool,
+    over: bool,
+}
+
+// Note that the operations are reversed because the exact approximation should be the top element in the lattice but we want it to be (false,false).
```

Comment:
I don't understand at all why it's this way. `over = true` should mean that e.g. `vp[S] >= wp[S]`. And the bitwise and implementation *must* be the obvious one, i.e. pairwise conjunction.

As I understand it now, `ApproximationKind` seems to be the *flipped* lattice of the one you put in the PR description.